go/cagra/index_params.go

```diff
@@ -0,0 +1,170 @@
+package cagra
+
+// #include <cuda_runtime_api.h>
+// #include <cuvs/core/c_api.h>
+// #include <cuvs/distance/pairwise_distance.h>
+// #include <cuvs/neighbors/brute_force.h>
+// #include <cuvs/neighbors/ivf_flat.h>
+// #include <cuvs/neighbors/cagra.h>
+// #include <cuvs/neighbors/ivf_pq.h>
+import "C"
+import (
+	"errors"
+	"unsafe"
+
+	cuvs "github.com/rapidsai/cuvs/go"
+)
+
+type IndexParams struct {
+	params C.cuvsCagraIndexParams_t
+}
+
+type CompressionParams struct {
+	params C.cuvsCagraCompressionParams_t
+}
+
+type BuildAlgo int
+
+const (
+	IvfPq BuildAlgo = iota
+	NnDescent
+	AutoSelect
+)
+
+var cBuildAlgos = map[BuildAlgo]int{
+	IvfPq:      C.IVF_PQ,
+	NnDescent:  C.NN_DESCENT,
+	AutoSelect: C.AUTO_SELECT,
+}
+
+func CreateCompressionParams() (*CompressionParams, error) {
+
+	size := unsafe.Sizeof(C.struct_cuvsCagraCompressionParams{})
+
+	params := (C.cuvsCagraCompressionParams_t)(C.malloc(C.size_t(size)))
+
+	if params == nil {
+		return nil, errors.New("memory allocation failed")
+	}
+
+	err := cuvs.CheckCuvs(cuvs.CuvsError(C.cuvsCagraCompressionParamsCreate(&params)))
+
+	if err != nil {
+		return nil, err
+	}
+
+	// params.pq_bits = C.uint32_t(pq_bits)
+	// params.pq_dim = C.uint32_t(pq_dim)
+	// params.vq_n_centers = C.uint32_t(vq_n_centers)
+	// params.kmeans_n_iters = C.uint32_t(kmeans_n_iters)
+	// params.vq_kmeans_trainset_fraction = C.double(vq_kmeans_trainset_fraction)
+	// params.pq_kmeans_trainset_fraction = C.double(pq_kmeans_trainset_fraction)
+
+	return &CompressionParams{params: params}, nil
+}
+
+func (p *CompressionParams) SetPQBits(pq_bits uint32) (*CompressionParams, error) {
+	p.params.pq_bits = C.uint32_t(pq_bits)
+	return p, nil
+}
+
+func (p *CompressionParams) SetPQDim(pq_dim uint32) (*CompressionParams, error) {
+	p.params.pq_dim = C.uint32_t(pq_dim)
+	return p, nil
+}
+
+func (p *CompressionParams) SetVQNCenters(vq_n_centers uint32) (*CompressionParams, error) {
+	p.params.vq_n_centers = C.uint32_t(vq_n_centers)
+	return p, nil
+}
+
+func (p *CompressionParams) SetKMeansNIters(kmeans_n_iters uint32) (*CompressionParams, error) {
+	p.params.kmeans_n_iters = C.uint32_t(kmeans_n_iters)
+	return p, nil
+}
+
+func (p *CompressionParams) SetVQKMeansTrainsetFraction(vq_kmeans_trainset_fraction float64) (*CompressionParams, error) {
+	p.params.vq_kmeans_trainset_fraction = C.double(vq_kmeans_trainset_fraction)
+	return p, nil
+}
+
+func (p *CompressionParams) SetPQKMeansTrainsetFraction(pq_kmeans_trainset_fraction float64) (*CompressionParams, error) {
+	p.params.pq_kmeans_trainset_fraction = C.double(pq_kmeans_trainset_fraction)
+	return p, nil
+}
+
+func CreateIndexParams() (*IndexParams, error) {
```

Comment:
Are these function names namespaced in some way? For example, will this conflict with an ivfpq `CreateIndexParams` or does the Go compiler know that this is different because it's nested in a `cagra` directory? I wonder if we should mirror the C nameing convention (i.e. cuvsCagraIndexParamsCreate, prefix functions with cuvs, followed by the subject and put the verb at the end).

Reply:
Nevermind, I see the `package cagra` at the top of the file so I'm assuming everything is nested inside. I think I'd prefer to fully qualify these packages, though, for consistency with the other cuVS APIs: `cuvs.neighbors.cagra`.